### DISCUSSION TITLE
CPU performance issues

### DISCUSSION THREAD

**Discussion by mihaho** [pointfreeco/swift-identified-collections#28] [Q&A]
I am facing some issues when trying to upgrade from Composable Architecture v20 to a higher version. The problem is that the CPU starts spiking after the introduction of the Identified Collections package. I profiled the app and noticed that the largest stack trace was due to the equality checks. This drew my attention to the equality conformance of `IdentifiedArray`. The difference with the old version is that you now use _[elementsEqual](https://github.com/apple/swift/blob/main/stdlib/public/core/SequenceAlgorithms.swift#L318-L336)_ instead of _[==](https://github.com/apple/swift/blob/main/stdlib/public/core/Array.swift#L1728-L1752)_. I  checked the difference in Swift's source code. _elementsEqual_ always creates 2 iterators for each sequence and goes through all elements while _==_ has some shortcuts.

Replacing `lhs.elementsEqual(rhs)` with `lhs.elements == rhs.elements` helps a lot. The app becomes responsive again and the CPU drops for about ~50%

Is _elementsEqual_ needed? Any comments are welcome

**Comment by mihaho** [comment]
As expected, the biggest difference is when we compare equal elements. I have run a benchmark test which is attached below. If you agree, I will be happy to create a pull request.

<img width="1280" alt="chart" src="https://user-images.githubusercontent.com/18087095/137743058-9995b3b3-d005-4881-a463-bead2c42e9b7.png">

Legend

```
Red line - current algorithm, where `elementsEqual` is replaced with `==`
Blue line - current algorithm, using `elementsEqual` (no changes)
```

Source code:

```Swift
import CollectionsBenchmark
import CollectionsA // red line
import CollectionsB // blue line

struct DummyState: Equatable, Identifiable {
    
    let value: Int
    
    var id: Int {
        value
    }
    
    init(_ value: Int) {
        self.value = value
    }
}

var benchmark = Benchmark(title: "Benchmark-Equality")

benchmark.add(
    title: "Equality ==",
    input: ([Int]).self
) { input in
    return { timer in
        let values = input.map(DummyState.init)
        let copiedValues = values
        
        let d1 = CollectionsA.IdentifiedArray(uniqueElements: values)
        let d2 = CollectionsA.IdentifiedArray(uniqueElements: copiedValues)
        
        timer.measure {
            blackHole(d1 == d2)
        }
    }
}
benchmark.add(
    title: "Equality elementsEqual",
    input: ([Int]).self
) { input in
    return { timer in
        let values = input.map(DummyState.init)
        let copiedValues = values
        
        let d1 = CollectionsB.IdentifiedArray(uniqueElements: values)
        let d2 = CollectionsB.IdentifiedArray(uniqueElements: copiedValues)
        
        timer.measure {
            blackHole(d1 == d2)
        }
    }
}

benchmark.main()
                                        
```

**Reply by stephencelis** [reply]
@mihaho Thanks for looking into this and for creating that benchmark! We'd definitely take a PR to change the implementation to call `==` instead. Wanna submit one? Otherwise we'll try to get to this soon.

**Comment by mihaho** [comment]
@stephencelis, thanks for answering. Sure, I'll submit a PR

**Comment by maximkrouk** [comment]
Also, I noticed that a thing mentioned in inline docs that direct access for `Identifiable.id` is faster than using a KeyPath, maybe it's better to add init with an _id closure to public API to enable users to specify id keypath as a function explicitly?
